python/ray/rllib/agent.py

```diff
@@ -25,6 +25,32 @@
 logger.setLevel(logging.INFO)
 
 
+def _deep_update(original, new_dict, new_keys_allowed, whitelist):
+    """Updates original dict with values from new_dict recursively.
+    If new key is introduced in new_dict, then if new_keys_allowed is not
+    True, an error will be thrown.
+
+    Args:
+        original (dict): Dictionary with default values.
+        new_dict (dict): Dictionary with values to be updated
+        new_keys_allowed (bool): Whether new keys are allowed.
+        whitelist (list): List of keys that will not be recursively
+            updated. This is only at the top level.
+    """
+    for k, value in new_dict.items():
+        if k not in original and k != "env":
+            if not new_keys_allowed:
+                raise Exception(
+                    "Unknown config parameter `{}` ".format(k))
+            else:
+                logger.warn("`{}` not in default configuration...".format(k))
+        if type(original.get(k)) is dict and k not in whitelist:
```

Comment:
Isn't this backwards? I want to be able to overwrite sub-fields of the tf config, it's just that I don't want that to throw "Unknown config param." I think the check for `k not in whitelist` should go when checking if `k not in original`.

Reply:
No, since the whitelist only applies at the top level.

```
whitelist = ['b']
default_config = {
    'a': 1,
    'b': { 'foob': 1, 'baz': 23}
    'c': { 'fooc': 2 }
}
new_config = {
    'b': { 'bar': 1 }
}
_deep_update(default_config, new_config, False, whitelist) == {'a': 1, 'b': { 'bar': 1 }, 'c': { 'fooc': 2 }}
new_config2 = {
    'c': { 'barc': 1 }
}
_deep_update(default_config, new_config2, False, whitelist) # throws an error

Reply:
Hmm, but what if I have
```
model {
  grayscale: True
}
```

and want to merge it with

```
model {
  layers: 4
}
```

(actually, I thought the original point of this PR was so that we could do deep merges of dictionaries).